keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -589,6 +618,83 @@ associativeList.all(e, e.key1 == 'a' && e.key2 == 'b' && e.val == 100)
 associativeList.all(e, e.val == 100)
 ```
 
+### Access to Existing Object State
+
+#### Use Cases
+
+| Use Case                                                          | CEL Rule                                                                        | Examples |
+| ----------------------------------------------------------------- | ------------------------------------------------------------------------------- | -------- |
+| Immutability                                                      | `self.foo == oldSelf.foo`                                                       ||
+| Prevent modification/removal once assigned                        | `oldSelf != 'bar' || self == 'bar'` or `!has(oldSelf.field) || has(self.field)` | [featureSet selection](https://github.com/openshift/kubernetes/blob/b426a3ffde5e7cd78483adeb5cb855f9f04dec5a/openshift-kube-apiserver/admission/customresourcevalidation/features/validate_features.go#L81-L88) |
+| Append-only set                                                   | `self.all(element, element in oldSelf)`                                         ||
+| If previous value was X, new value can only be A or B, not Y or Z | `oldSelf != 'X' || self in ['A', 'B']`                                          | [Latency profile selection](https://github.com/openshift/enhancements/pull/977) |
+| Tolerating once-acceptable values ("ratcheting")                  | `self == oldSelf || self in ["currently", "acceptable", "values"]`              ||
+
+#### Design Considerations
+
+- The use of "old" in the identifier "oldSelf" is congruent with how `AdmissionReview` identifies
+  the existing object as `oldObject`.
+
+- CEL doesn't support checking if a root variable is bound. The macro `has` can test for the presence of fields, as in `has(self.child)`, but `has(oldSelf)` is not legal.
+
+- Possible interactions with [immutability and validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
+
+#### Alternatives
+
+1. A boolean field is added to the extension indicating when true that the rule applies only to
+   updates. The expression in "rule" may not reference the identifier "oldSelf" unless "updateRule"
+   is present and true. If "oldSelf" is referenced by the expression, and updateRule is absent or
+   false, an error is generated at CRD update time.
+
+```yaml
+x-kubernetes-validations:
+  - rule: "oldSelf.xyz"
+    updateRule: true
+```
+
+2. Allowing rules to specify their expression in one string field named either "rule" or
+   "updateRule" is effectively equivalent to the first example. The decision to populate
+   "updateRule" instead of "rule" is treated as though the boolean field were true in option 1.
+
+```yaml
+x-kubernetes-validations:
+  - updateRule: "oldSelf.xyz"
+```
+
+3. Any reference within the "rule" expression to the identifier "oldSelf" could implicitly mark the
```

Comment:
@liggitt  would this resolve your concern about constraining future choice by not making a new category of rule?

I like this option because I think it's easier for people writing rules.